src/auth/src/token_cache.rs

```diff
@@ -0,0 +1,351 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::token::{Token, TokenProvider};
+use crate::Result;
+use std::sync::Arc;
+use tokio::sync::{Mutex, Notify};
+// Using tokio's wrapper makes the cache testable without relying on clock times.
+use tokio::time::Instant;
+
+// TODO(#1210) - implementation in progress
+#[allow(dead_code)]
+#[derive(Debug)]
+pub(crate) struct TokenCache<T>
+where
+    T: TokenProvider,
+{
+    // The cached token, or the last seen error.
+    token: Arc<Mutex<Result<Token>>>,
+
+    // Tracks if a refresh is ongoing. If the lock is held, there is a refresh.
+    refresh_in_progress: Arc<Mutex<()>>,
+    // Allows us to await the result of a refresh in multiple threads.
+    refresh_notify: Arc<Notify>,
+
+    // The token provider. This thing does the refreshing.
+    inner: Arc<T>,
+}
+
+// Returns true if we are holding an error, or a token that has expired.
+fn invalid(token: &Result<Token>) -> bool {
+    match token {
+        Ok(t) => t.expires_at.is_some_and(|e| e <= Instant::now().into_std()),
+        Err(_) => true,
+    }
+}
+
+// We manually implement the `Clone` trait because the Rust compiler will
+// squawk if `T` is not `Clone`, even though we only hold an `Arc<T>`. :shrug:
+impl<T: TokenProvider> Clone for TokenCache<T> {
+    fn clone(&self) -> TokenCache<T> {
+        TokenCache {
+            token: self.token.clone(),
+            refresh_in_progress: self.refresh_in_progress.clone(),
```

Comment:
is the task doing the refresh also cloned? If not, this cache will permanently be in a refreshing state (if cloned at the time when a refresh was happening)

Reply:
We are cloning the `Arc`. The stuff in the `Arc` doesn't change.

Reply:
So lets say thread A is doing refresh and has a lock on the mutex in refresh_in_progress. A clone is called when this is happening, what will be the state of the mutex in the clone?

Reply:
It would still be locked in the clone. Clones of `Arc<Mutex<T>>` just share the underlying `Mutex<T>`.

Note that `tokio::sync::Mutex` cannot be cloned: https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html#trait-implementations.